chat_platform.py

```diff
@@ -0,0 +1,212 @@
+
+    def send_message(self, channel: str, text: str, thread_id: Optional[str] = None):
+        self.slack_app.client.chat_postMessage(channel=channel, text=text, thread_ts=thread_id)
+
+    def add_reaction(self, channel: str, name: str, timestamp: str):
+        self.slack_app.client.reactions_add(channel=channel, name=name, timestamp=timestamp)
+
+    def remove_reaction(self, channel: str, name: str, timestamp: str):
+        self.slack_app.client.reactions_remove(channel=channel, name=name, timestamp=timestamp)
+
+    @lru_cache(maxsize=30)
+    def get_message(self, channel: str, event_id: str) -> str:
+        timestamp = event_id
+        result = self.slack_app.client.conversations_history(
+            channel=channel,
+            latest=timestamp,
+            inclusive=True,
+            limit=1
+        )
+        return result["messages"][0]["text"]
+
+    def handle_app_mention_events(self, body, logger, say):
+        logger.info(body)
+        event = body["event"]
+        channel = event["channel"]
+        thread_id = event.get("thread_ts", None)
+        event_id = event["ts"]
+        self.add_reaction(channel=channel, name='hourglass_flowing_sand', timestamp=event_id)
+        result = self.process_message_func(
+            channel=channel,
+            event_id=event_id,
+            thread_id=thread_id,
+            text=event["text"],
+        )
+        logging.info(f"Result from AI: {result}")
+        say(text=result.output, thread_ts=thread_id or event_id)
+        slack_app.client.reactions_remove(channel=channel, name='hourglass_flowing_sand', timestamp=event_id)
+
+
+class MatrixPlatform(ChatPlatform):
+    def __init__(self, homeserver_url: str, access_token: str,
+            matrix_room: str, process_message_func):
+        self.starttime = int(time.time() * 1000) # current time in milliseconds
+        self.homeserver_url = homeserver_url
+        self.client = AsyncClient(
+            homeserver=homeserver_url,
+            store_path="my_matrix_store",
+            config=ClientConfig(
+                store_sync_tokens=True,
+                store=SqliteStore
+            )
+        )
+        self.client.access_token = access_token
+        asyncio.run(self.client.whoami()) # updates user_id device_id
+        self.client.user = self.client.user_id
+        self.bare_username = self.client.user_id[1:].split(":")[0]
+        self.matrix_room = matrix_room
+        self.process_message_func = process_message_func
+        response = asyncio.run(self.client.room_resolve_alias(self.matrix_room))
+        self.matrix_room_id = response.room_id
+        self.client.load_store()
+        logging.info(f"XXX: {self.client.user_id}")
+
+    def send_message(self, channel: str, text: str, thread_id: Optional[str] = None):
+        print(f"XXX: send_message channel: {channel}")
+        print(f"XXX: send_message thread_id: {thread_id}")
+        print(f"XXX: send_message text: {text}")
+        try:
```

Comment:
We should probably be consistent here wrt error handling between the Slack and Matrix backends. Probably makes sense to add try/except blocks to Slack so it doesn't crash the app entirely if there's a network flake.